src/provider/database.rs

```diff
@@ -6,21 +6,880 @@ use crate::{
+            Either::Left(data) => data.get_ref().clone(),
+            Either::Right(mut file) => {
+                let mut data = Vec::new();
+                file.read_to_end(&mut data).await?;
+                Bytes::from(data)
+            }
+        })
+    }
+
+    pub fn into_inner(self) -> Either<std::io::Cursor<Bytes>, tokio::fs::File> {
+        self.0
+    }
+}
+
+impl tokio::io::AsyncRead for AsyncDataReader {
+    fn poll_read(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<tokio::io::Result<()>> {
+        Pin::new(&mut self.0).poll_read(cx, buf)
+    }
+}
+
+impl tokio::io::AsyncSeek for AsyncDataReader {
+    fn start_seek(mut self: std::pin::Pin<&mut Self>, position: io::SeekFrom) -> io::Result<()> {
+        Pin::new(&mut self.0).start_seek(position)
+    }
+
+    fn poll_complete(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<io::Result<u64>> {
+        Pin::new(&mut self.0).poll_complete(cx)
+    }
+}
+
+impl BlobData {
+    fn reader(&self, default_path: impl Fn() -> PathBuf) -> io::Result<DataReader> {
+        Ok(match self {
+            Self::Inline(data) => DataReader::Memory(std::io::Cursor::new(data.clone())),
+            Self::File(custom_path) => DataReader::File(match custom_path {
+                Some(path) => std::fs::File::open(path)?,
+                None => std::fs::File::open(default_path())?,
+            }),
+        })
+    }
+
+    fn async_reader(
+        &self,
+        default_path: impl Fn() -> PathBuf,
+    ) -> impl Future<Output = io::Result<AsyncDataReader>> + 'static {
+        match self {
+            Self::Inline(data) => {
+                futures::future::ready(Ok(AsyncDataReader::memory(data.clone()))).left_future()
+            }
+            Self::File(custom_path) => {
+                let path = match custom_path {
+                    Some(path) => path.clone(),
+                    None => default_path(),
+                };
+                async move {
+                    let file = tokio::fs::File::open(path).await?;
+                    Ok(AsyncDataReader(Either::Right(file)))
+                }
+                .right_future()
+            }
+        }
+    }
+
+    fn exists(&self, default_path: impl Fn() -> PathBuf) -> bool {
+        match self {
+            Self::Inline(_) => true,
+            Self::File(Some(path)) => path.exists(),
+            Self::File(None) => default_path().exists(),
+        }
+    }
+}
+
+/// A bao file consists of outboard and content.
+///
+/// Either of them can stored inline or in a file in the file system.
+///
+/// Under some circumstances, we can have multiple content files. We will
+/// however not keep track of multiple outboard files.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct BaoFile {
+    outboard: BlobData,
+    content: Vec<BlobData>,
```

Comment:
I'm a little unhappy this leaves the possibility for the raw data in memory multiple times.  Though it's behind a `Bytes` which has an `Arc` built-in, but still.  Maybe this `Vec` is better pushed down to a type distinguishing the in-mem storage from the file storage.

Reply:
I think I am deduping this somewhere, but you are right that this could be hidden in its own type. I just want to keep these things simple since they also define the serialisation format.